### DISCUSSION TITLE
Get all nodes from sever.

### DISCUSSION THREAD

**Discussion by Yoshibeatsu** [FreeOpcUa/python-opcua#1543] [Q&A]
Hi, it might be not common practice but i need to get all the opc nodes from one server. I couldn't find an example which is getting all the nodes and also is using certificates and user login as well. I'm quite new to programming especially python. The nodes are in folders and sub sub folders. So I think i need to make a recursive search ...

I tried already to develop something in C# with the libary by the opc foundation. It worked on my develop system but ran into a timeout after a while on a real opc server.... 

Thank you guys in advance

**Comment by oroulet** [comment]
You got a timeout because trying to get all nodes does not make any sense. Try to search the net, you are not the first one to try

**Reply by Yoshibeatsu** [reply]
I got this request from my project manager. It wasn't my idea. I already search the net and couldn't find a good way to solve this issue. Might be easier for you sir to solve this issue but I'm quite new at programming as I said. Sorry for asking a question which seems to be obsolete for you but this is a real problem for me

**Comment by phillipp-schlueter** [comment]
I'm not certain weather this approach is what your looking for but I'm using the following recursive function to loop through all elements of a server to display them as a Treeview
```
client = Client("opc.tcp://localhost:4840/freeopcua/server/")
client.connect()
root = client.get_root_node()
get_server_tree_recursive(client, root)

def get_server_tree_recursive(client, parent):
    children = client.get_node(parent).get_children()
    if children:
        for child in children:
            child_id = client.get_node(child)
            parent_id = client.get_node(parent)
            print(child_id.get_display_name()._text ' is child of ', parent_id.get_display_name()._text)
            get_server_tree_recursive(client, child)
    else:
        return 0
```

**Reply by AndreasHeine** [reply]
actually the tools like uaexpert do it on click not upfront so if you want to unfold a node you browse it onclick

**Reply by phillipp-schlueter** [reply]
Yeah I'm aware but for the program I'm writing I need the user to be able to search the server just like uaexpert displays the content

**Reply by AndreasHeine** [reply]
loading all nodes from root in memory is a bad idea... it containes types and normative stuff its like million's of nodes (on servers i work on)
a better aproach is to make a depth first search following just the nodeid's (as string) instead of storing all Node-Class instances and apply filters for what your looking for

edit: its like storing just the cursor where your in the tree

**Reply by AndreasHeine** [reply]
remember you can browse up (inverse) and down!

**Reply by phillipp-schlueter** [reply]
I can't quit follow what you criticize. In my understanding I'm working through the nodes step by step, only having its parents loaded (which mostly are not millions). I'm only printing or in my case storing their display names not the whole node-class instance.
Could you maybe specify which part of the code concerns you?

**Reply by AndreasHeine** [reply]
![image](https://github.com/user-attachments/assets/6748f9b6-2444-4063-a53e-3f271eaecea2)

Root childs are:
-Objects (kind of a better startpoint because it contains instances rather then definitions/declarations)
-Types (normative nodes from spec. or "companion specs" / "XML imports" which are the largest part of the nodes on a server)
-Views (might contain duplicates to Objects recursive childs)

**Reply by AndreasHeine** [reply]
your code creates a lot of requests in a short period of time! and your not doing anything with the stuff you found even if you build a tree to display (you need to store the result somewhere) which need a lot of memory... like i mentioned UaExpert shows you a tree as well but it uses the click and select events to browse/read thru the addressspace while the user go thru the tree

**Reply by phillipp-schlueter** [reply]
I'm really sorry but I don't get what you're trying to say.
